### PR TITLE
Support exception types for >= error do-blocks

### LINKED ISSUE #127: Allow `Expection` in do-block syntax
Do-block sytax can be used for `String`s
```julia
julia> error(getlogger()) do
           "Woops!"
       end
[error | root]: Woops!
ERROR: Woops!
Stacktrace:
 [1] error(::getfield(Main, Symbol("##21#22")), ::Logger) at /Users/nick/.julia/packages/Memento/UgJr2/src/loggers.jl:434
 [2] top-level scope at none:0
```

But does not work for `Excpetion`s
```julia
julia> error(getlogger()) do
           DomainError("Woops!")
       end
ERROR: MethodError: Cannot `convert` an object of type DomainError to an object of type AbstractString
```

Would it be possible for this to be added?

### PR DESCRIPTION
**Pull request by rofinn**
Closes #127

### CODE REVIEW

File: src/loggers.jl [hunk 1]
```diff
@@ -430,8 +430,18 @@ let
...
 
                     function $level(msg::Function, logger::Logger)
-                        log(msg, logger, $key)
-                        throw(ErrorException(msg()))
+                        # NOTE: We can't lazily evaluate do-blocks on errors or higher
+                        # because we need to determine whether the function returns
+                        # a string or an error type. We provide this syntax for readability
+                        # rather than performance.
+                        m = msg()
+                        if isa(m, String)
```
Comment (omus) [inline src/loggers.jl]:
```suggestion
                        if m isa AbstractString
```

File: src/loggers.jl [hunk 2]
```diff
@@ -430,8 +430,18 @@ let
...
+                            log(logger, $key, m)
+                            throw(ErrorException(m))
+                        else
+                            log(logger, $key, sprint(io -> showerror(io, m)))
```
Comment (omus) [inline src/loggers.jl]:
```suggestion
                            log(logger, $key, sprint(showerror, m))
```

File: src/loggers.jl [hunk 3]
```diff
@@ -430,8 +430,18 @@ let
...
+                            throw(m)
+                        end
```
Comment (omus) [inline src/loggers.jl]: I think you may want to check `m isa Exception` instead as someone may return a number or some other type which can be stringified.

File: test/loggers.jl [hunk 4]
```diff
@@ -82,6 +82,21 @@
...
+                    string(
+                        "sqrt will only return a complex result if called with a complex ",
+                        "argument. Try sqrt(Complex(x))."
+                    ),
+                )
+            end
+
+            @test_throws ErrorException Memento.error(logger) do
+                "I'm an error msg."
+            end
```
Comment (omus) [inline test/loggers.jl]: Do you also want to test that these emitted logs?

File: src/loggers.jl (hunk 1 above)
Comment (rofinn) [reply src/loggers.jl]: Not a fan of the infix syntax

File: src/loggers.jl (hunk 3 above)
Comment (rofinn) [reply src/loggers.jl]: So you're thinking that we should fall back to string `string(x)`?

File: test/loggers.jl (hunk 4 above)
Comment (rofinn) [reply test/loggers.jl]: Yeah, probably worth testing.

File: src/loggers.jl (hunk 1 above)
Comment (omus) [reply src/loggers.jl]: That's fine. Should still use `AbstractString`
```suggestion
                        if isa(m, AbstractString)
```

File: src/loggers.jl (hunk 3 above)
Comment (omus) [reply src/loggers.jl]: Yes, mainly I was thinking about:
```julia
if isa(m, Exception)
    ...
else
    ...
end
```

Comment (rofinn) [reply src/loggers.jl]: 👌

**Pull request merged**